File: Task/Search-a-list-of-records/Lua/search-a-list-of-records.lua
-- Dataset declaration
local cityPops = {
    {name = "Lagos", population = 21.0},
    {name = "Cairo", population = 15.2},
    {name = "Kinshasa-Brazzaville", population = 11.3},
    {name = "Greater Johannesburg", population = 7.55},
    {name = "Mogadishu", population = 5.85},
    {name = "Khartoum-Omdurman", population = 4.98},
    {name = "Dar Es Salaam", population = 4.7},
    {name = "Alexandria", population = 4.58},
    {name = "Abidjan", population = 4.4},
    {name = "Casablanca", population = 3.98}
}

-- Function to search a dataset using a custom match function
function recordSearch (dataset, matchFunction)
    local returnValue
    for index, element in pairs(dataset) do
        returnValue = matchFunction(index, element)
        if returnValue then return returnValue end
    end
    return nil
end

-- Main procedure
local testCases = {
    function (i, e) if e.name == "Dar Es Salaam" then return i - 1 end end,
    function (i, e) if e.population < 5 then return e.name end end,
    function (i, e) if e.name:sub(1, 1) == "A" then return e.population end end
}
for _, func in pairs(testCases) do print(recordSearch(cityPops, func)) end

File: Task/Search-a-list-of-records/Free-Pascal-Lazarus/search-a-list-of-records.pas
program Search_list_records;
{$mode ObjFPC}{$H+}

type
  TCity = record
    name: string;
    population: real;
  end;

const
  Cities: array of TCity = (
    (name: 'Lagos'; population: 21.0),
    (name: 'Cairo'; population: 15.2),
    (name: 'Kinshasa-Brazzaville'; population: 11.3),
    (name: 'Greater Johannesburg'; population: 7.55),
    (name: 'Mogadishu'; population: 5.85),
    (name: 'Khartoum-Omdurman'; population: 4.98),
    (name: 'Dar Es Salaam'; population: 4.7),
    (name: 'Alexandria'; population: 4.58),
    (name: 'Abidjan'; population: 4.4),
    (name: 'Casablanca'; population: 3.98)
  );

function FindCityIndex(const CityName: string): Integer;
var
  i: Integer;
begin
  Result := -1;
  for i := 0 to High(Cities) do
    if Cities[i].name = CityName then
      Exit(i);
end;

function FindCityName(const pop: real): string;
var
  City: TCity;
begin
  Result := 'not found';
  for City in Cities do
    if City.population < pop then
      Exit(City.name);
end;

function FindCityPopulation(const Start: Char): Real;
var
  City: TCity;
begin
  Result := -1;
  for City in Cities do
    if City.name[1] = Start then
      Exit(City.population);
end;

begin
  writeln('index: ', FindCityIndex('Dar Es Salaam'));
  writeln('name: ', FindCityName(5.0));
  writeln('population: ', FindCityPopulation('A'):4:2);
end.
